### PR TITLE
[fix][ML]Fix NPE when put value to `RangeCache`.

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

When `ReferenceCounted` object overrides the method `deallocate` to make the `getLength` value equal null will cause NPE because the `RangeCache#put` method is not thread-safe. (The process of describing this abstraction is not very clear, please refer to the code modification :)

Pulsar implementation may throw an exception to make `OpAddEntry` fail abnormal and fence the topic. relative code as below:

https://github.com/apache/pulsar/blob/defeec0e84a63ea865f3a2790bc61b66a02254c5/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java#L211-L217

**Exception screenshot:**

```java
java.lang.NullPointerException: Cannot invoke "String.length()" because "value.s" is null

	at org.apache.bookkeeper.mledger.util.RangeCacheTest.lambda$testInParallel$6(RangeCacheTest.java:279)
	at org.apache.bookkeeper.mledger.util.RangeCache.put(RangeCache.java:77)
	at org.apache.bookkeeper.mledger.util.RangeCacheTest.testInParallel(RangeCacheTest.java:283)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
```

### Modifications

- Make the `RangeCache#put` method to thread-safe.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]
Nice catch!

**Review by Technoboy- — APPROVED** [review]

**Comment by Technoboy-** [comment]
This may be a core cause of topic fenced, but we are still confirming.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java [hunk 1]
```diff
@@ -73,12 +73,11 @@ public RangeCache(Weighter<Value> weighter, TimestampExtractor<Value> timestampE
...
-        if (entries.putIfAbsent(key, value) == null) {
+        Value v = entries.computeIfAbsent(key, (k) -> {
             size.addAndGet(weighter.getSize(value));
-            return true;
-        } else {
-            return false;
-        }
+            return value;
+        });
+        return v == value;
```
Comment (Demogorgon314) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java:80]: Nice catch! But look like it changed the original behavior. See the following test. Its success in the old code but failed in this change.
```java
    @Test
    public void test() {
        RangeCache<Integer, RefString> cache = new RangeCache<>();
        RefString s0 = new RefString("zero");
        assertEquals(s0.refCnt(), 1);
        assertTrue(cache.put(0, s0));
        assertFalse(cache.put(0, s0));
    }
```

Comment (mattisonchao) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java]: Got it, very thanks to you.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java [hunk 2]
```diff
@@ -73,12 +74,13 @@ public RangeCache(Weighter<Value> weighter, TimestampExtractor<Value> timestampE
      * @return whether the entry was inserted in the cache
      */
     public boolean put(Key key, Value value) {
-        if (entries.putIfAbsent(key, value) == null) {
+        AtomicBoolean flag = new AtomicBoolean(false);
```
Comment (Demogorgon314) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java:77]: Better be using `MutableBoolean`?

**Review by Demogorgon314 — APPROVED** [review]

**Review by nodece — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Review by BewareMyPower — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java [hunk 3]
```diff
@@ -73,12 +74,13 @@ public RangeCache(Weighter<Value> weighter, TimestampExtractor<Value> timestampE
...
+        entries.computeIfAbsent(key, (k) -> {
             size.addAndGet(weighter.getSize(value));
-            return true;
-        } else {
-            return false;
-        }
+            flag.setValue(true);
+            return value;
+        });
+        return flag.booleanValue();
```
[the remark this answers is not in the document]
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java:83]: @mattisonchao  issue is #21301

[the remark this answers is not in the document]
Comment (mattisonchao) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java:83]: Yes, you are right.

**Pull request merged**